middle_end/flambda2/simplify/simplify_let_cont_expr.ml

```diff
@@ -1482,10 +1830,60 @@ let simplify_let_cont0 ~(simplify_expr : _ Simplify_common.expr_simplifier) dacc
     DE.add_lifting_cost lifting_cost
       (DE.define_continuations denv_for_body bound_continuations)
   in
+  (* During specialization, we must take care of correctly handling let-bound
+     continuations that have been lifted during the first downwards pass, and
+     whhose handlers must be replaced on subsequent passes, so that they can
+     refer to the lifted ones. *)
+  if debug ()
+  then
+    Format.eprintf "--- LET_CONT0@\n%a@\n%a@\n@." Replay_history.print
+      (DE.replay_history denv_for_body)
+      Original_handlers.print data.handlers;
+  let handlers =
+    match
+      Replay_history.replay_continuation_mapping
+        (DE.replay_history denv_for_body)
+    with
+    | Still_recording -> data.handlers
+    | Replayed continuation_mapping -> (
+      match data.handlers with
+      | Non_recursive non_rec_handler ->
+        if debug ()
+        then
+          Format.eprintf "+++ LIFTED ? %a@\nreplay: %a@\n@." Continuation.print
+            non_rec_handler.cont Replay_history.print
+            (DE.replay_history denv_for_body);
+        let lifted_cont =
+          Continuation.Map.find non_rec_handler.cont continuation_mapping
+        in
+        let new_handler =
+          let args =
+            List.map
+              (fun bp -> Simple.var (Bound_parameter.var bp))
+              (Bound_parameters.to_list non_rec_handler.params)
+          in
+          let apply_cont =
+            Apply_cont.create lifted_cont ~dbg:Debuginfo.none ~args
+          in
+          Flambda.Expr.create_apply_cont apply_cont
+        in
+        if debug ()
+        then
+          Format.eprintf "*** LIFTED %a:@\n%a@\n@." Continuation.print
+            non_rec_handler.cont Flambda.Expr.print new_handler;
+        let new_non_rec_handler =
+          Non_recursive_handler.with_handler new_handler non_rec_handler
+        in
+        Original_handlers.create_non_recursive new_non_rec_handler
+      | Recursive _ ->
+        (* TODO: handle these *)
+        assert false)
```

Comment:
I think we don't specialize recursive continuations, we should at least put a nice `Misc.fatal_error` here